### ISSUE TITLE
Define custom positon for `qml.transforms.insert`

### ISSUE THREAD

**Issue by ankit27kh** [PennyLaneAI/pennylane#2413]
### Feature details

Currently, the insert transform supports three position settings, `end`, `all` and `start`. `all` and `start` will ignore state preparations.
But this only applies to the templates already present in PennyLane. But we can have custom state preparation routines. It will be nice to have some way to mark our state preparation which is then not modified with the insert transform.

Also, only state preparations should be skipped, which come at the beginning of the circuit. Currently, if I add another template like `StronglyEntanglingLayers` somewhere in between, it is also skipped over by the transform. I don't think this is the expected behaviour.

### Implementation

State preparations can be asked to be made as a separate function (like a template) outside the circuit and then called from within. These will then have a specific argument or name instructing the transform to skip.
A named argument seems like the better way. This can also be added to all templates. Users can then control if they want the transform to skip the template or not by marking it True or False.

### How important would you say this feature is?

1: Not important. Would be nice to have.

**Comment by josh146** [comment]
Thanks for reporting this @ankit27kh! This sounds like two separate issues; a bug report and a feature request.

Do you mind posting two separate code snippets?

1. One showing the bug where templates such as `StronglyEntanglingLayers` are skipped by the transform, and
2. One showing the feature request (e.g., marking the custom state preparation?)

This will help us debug and also plan enhancements 🙂

**Comment by ankit27kh** [comment]
Hey @josh146, I didn't know it was a bug! Here are separate code snippets:
1. Skipped template by the transform:
```
dev = qml.device("default.mixed", wires=2)

shape = qml.StronglyEntanglingLayers.shape(n_layers=2, n_wires=2)
weights = np.random.random(size=shape)

@qml.qnode(dev)
@qml.transforms.insert(qml.BitFlip, 1, position="all")
def f(w, x):
    qml.RX(w, wires=0)
    qml.RY(x, wires=1)
    qml.StronglyEntanglingLayers(weights=weights, wires=range(2))
    return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))

print(qml.draw(f)(0.9, 0.4))
>>> 0: ──RX(0.90)──BitFlip(1.00)─╭StronglyEntanglingLayers(M0)──BitFlip(1.00)─┤ ╭<Z@Z>
>>> 1: ──RY(0.40)──BitFlip(1.00)─╰StronglyEntanglingLayers(M0)──BitFlip(1.00)─┤ ╰<Z@Z>

print(qml.draw(f, expansion_strategy="device")(0.9, 0.4))
>>> 0: ──RX(0.90)──BitFlip(1.00)──Rot(0.98,1.00,0.18)─╭C─╭X──Rot(0.99,0.95,0.28)─╭C─╭X──BitFlip(1.00)─┤ ╭<Z@Z>
>>> 1: ──RY(0.40)──BitFlip(1.00)──Rot(0.71,0.74,0.97)─╰X─╰C──Rot(0.61,0.78,0.30)─╰X─╰C──BitFlip(1.00)─┤ ╰<Z@Z>
```
In this, we can see, that the inserted gate is only after the complete template and not propagating inside it.

2. Feature request:
```
dev = qml.device("default.mixed", wires=2)

@qml.qnode(dev)
@qml.transforms.insert(qml.BitFlip, 1, position="all")
def f(w, x):
    qml.MottonenStatePreparation(a, wires=range(2))
    qml.RX(w, wires=0)
    qml.RY(x, wires=1)
    return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))

print(qml.draw(f)(0.9, 0.4))
>>> 0: ─╭MottonenStatePreparation(M0)──BitFlip(1.00)──RX(0.90)──BitFlip(1.00)─┤ ╭<Z@Z>
>>> 1: ─╰MottonenStatePreparation(M0)──BitFlip(1.00)──RY(0.40)──BitFlip(1.00)─┤ ╰<Z@Z>
print(qml.draw(f, expansion_strategy="device")(0.9, 0.4))
>>> 0: ──RY(2.30)─╭C───────────╭C──BitFlip(1.00)──RX(0.90)──BitFlip(1.00)─┤ ╭<Z@Z>
>>> 1: ──RY(2.03)─╰X──RY(0.18)─╰X──BitFlip(1.00)──RY(0.40)──BitFlip(1.00)─┤ ╰<Z@Z>
```
Here, we see that the state preparation is skipped as it should. But, if we have custom state prep for example:
```
dev = qml.device("default.mixed", wires=2)

@qml.qnode(dev)
@qml.transforms.insert(qml.BitFlip, 1, position="all")
def f(w, x):
    state_prep()
    qml.RX(w, wires=0)
    qml.RY(x, wires=1)
    return qml.expval(qml.PauliZ(0) @ qml.PauliZ(1))

def state_prep():
    qml.CNOT(wires=[0, 1])
    qml.RY(1, wires=0)
    qml.RX(2, wires=1)
    qml.RY(1, wires=1)
    qml.RX(2, wires=1)

print(qml.draw(f)(0.9, 0.4))
>>> 0: ─╭C──BitFlip(1.00)──RY(1.00)──BitFlip(1.00)──RX(0.90)──BitFlip(1.00)────────────────────────────────────────────────────┤ ╭<Z@Z>
>>> 1: ─╰X──BitFlip(1.00)──RX(2.00)──BitFlip(1.00)──RY(1.00)──BitFlip(1.00)──RX(2.00)──BitFlip(1.00)───RY(0.40)──BitFlip(1.00)─┤ ╰<Z@Z>
```
Here, the transform has inserted the gates in my state preparation circuit. This should be user-controllable by having a keyword argument like `insert=False`. The current templates can then also be updated to have this argument to control whether we want the transform to insert inside or not.

**Comment by ankit27kh** [comment]
Hi @trbromley,

> For your second point, I want to note that `MottonenStatePreparation` is _not_ being skipped (you can see the following `BitFlip` operations).

This is what I meant by skipping. The `BitFlip` operations are *after* the template gates and not *inside* the template.
You can see the difference in the next example. There the `BitFlip` gates are after every gate of the template.

**Comment by CatalinaAlbornoz** [comment]
Thanks for clarifying this @ankit27kh!
